### ISSUE TITLE
Drop `rust-analyzer.cargo.sysrootQueryMetadata` and support for stitched sysroot workspace

### ISSUE THREAD

**Issue by Veykril** [rust-lang/rust-analyzer#18590]
https://github.com/rust-lang/rust-analyzer/pull/18511 added the config back, ultimately we should get rid of the entire stitched concept as we won't need it anymore for newer sysroots.

**Comment by Veykril** [comment]
> Maybe it's specific to json workspace project loading? I'll try to get a simple repro today

_Originally posted by @darichey in https://github.com/rust-lang/rust-analyzer/issues/18511#issuecomment-2511385319_
   
The json part might be it, or rather your environment. I assume your env might not support running cargo as is for the sysroot discovery?

**Comment by lnicola** [comment]
CC https://github.com/rust-lang/rust-analyzer/issues/17759#issuecomment-2477844162.

**Comment by davidbarsky** [comment]
Yeah, the default development environment we operate in has an outbound firewall to the entire internet and requires deliberate, per-program actions to opt-out. Additionally, cargo is not preinstalled on devservers, Buck is assumed to be the default.

(We do have a buckified sysroot, though...)

**Comment by Veykril** [comment]
> Additionally, cargo is not preinstalled on devservers, Buck is assumed to be the default. 

That is the issue then, although we should fall back to the stitched one in that case I feel like. I guess my recovery logic fails there (when the `--no-deps` call still doesnt yield results).

That does raise a good point though, we can't assume cargo exists for people using the json format (I do wanna get rid of the current stitched system either way so there is something to be designed here that replaces that)

**Comment by darichey** [comment]
> > Additionally, cargo is not preinstalled on devservers, Buck is assumed to be the default.
> 
> That is the issue then

I don't think this is it! The problem only occurred for people who _do_ have cargo installed!

**Comment by Veykril** [comment]
That is even more perplexing 😬 though i still think the approach for replacing the stitched one stands nevertheless (which would solve it either way here, or well, should)

**Comment by darichey** [comment]
@Veykril I can repro with only this patch to force `--no-deps` when loading the sysroot: https://gist.github.com/darichey/f3173b82a2eaec75473419d0cc661f93. 

@davidbarsky Pointed out that the "use X instead of cargo for sysroot metadata" setting should be independent of rust-project.json/discoverCommand because most users of Buck/Bazel will still want to use Cargo for the sysroot. I'm happy to work on this if you agree.

**Comment by Veykril** [comment]
Yes that sounds fine to me. So always try using cargo for cargo workspaces, but for rust-project.json the json either specifies the sysroot layout manually (it stitches it itself, or sets it empty if it doesn't want a sysroot) or it doesn't in which case we shell out to cargo. That sounds reasonable to me

**Comment by Veykril** [comment]
https://github.com/rust-lang/rust-analyzer/pull/18788 drops the config again but disables running cargo entirely for rust-project projects.
